src/pagination.rs

```diff
@@ -212,10 +256,80 @@ fn sort_posts(posts: &mut Vec<&liquid::value::Value>, config: &PaginationConfig)
     })
 }
 
+fn distribute_posts_by_tags<'a>(
+    all_posts: &[&'a liquid::value::Value],
+) -> HashMap<String, Vec<&'a liquid::value::Value>> {
```

Comment:
If we're already borrowing the values, should we borrow the keys?

Reply:
I've tried to borrow the keys as well:
```rust
fn distribute_posts_by_tags<'a>(
    all_posts: &[&'a liquid::value::Value],
) -> Result<HashMap<&'a str, Vec<&'a liquid::value::Value>>> {
    let mut per_tags: HashMap<&'a str, Vec<&'a liquid::value::Value>> = HashMap::new();
    let add_to_tag =
        |tag: &'a str,
         post: &'a liquid::value::Value,
         per_tags: &mut HashMap<&'a str, Vec<&'a liquid::value::Value>>| {
            let cur_tag = per_tags.entry(tag).or_insert(vec![post]);
            cur_tag.push(post)
        };

    for post in all_posts {
        if let Some(tags) = extract_tags(post) {
            for tag in tags {
                let tag = tag.as_scalar().ok_or("Should have string tags")?.to_str();
                add_to_tag(&tag, post, &mut per_tags)
            }
        } else {
            add_to_tag("_none", post, &mut per_tags)
        }
    }
    Ok(per_tags)
}
```
but I'm having lifetime errors:

```rust
error[E0597]: `tag` does not live long enough
   --> src\pagination.rs:276:29
    |
276 |                 add_to_tag(&tag, post, &mut per_tags)
    |                             ^^^ borrowed value does not live long enough
277 |             }
    |             - borrowed value only lives until here
    |
note: borrowed value must be valid for the lifetime 'a as defined on the function body at 260:29...
   --> src\pagination.rs:260:29
    |
260 | fn distribute_posts_by_tags<'a>(
    |                             ^^
```

Reply:
```rust
                let tag = tag.as_scalar().ok_or("Should have string tags")?.to_str();
                add_to_tag(&tag, post, &mut per_tags)
```

Oh, the problem is the `to_str`.  It might create a String which lives to the end of the scope rather than for lifetime `'a`.  You'd have to pattern match and pull out the `Borrowed` and error if you get an `Owned` (since we define tags as only strings).

This will be improved as we move tags to the frontmatter because then the code will only accept `Vec<String>`.

Man, took me a good number of passes to spot where the lifetime mismatch was.

Reply:
At least, you got it, I'm still quite lost every time lifetime is involved x)

Reply:
Now that tags are guaranteed to be strings, can't we switch to borrows?

Reply:
I tried:
```rust
fn distribute_posts_by_tags<'a>(
    all_posts: &[&'a liquid::value::Value],
) -> Result<HashMap<&'a str, Vec<&'a liquid::value::Value>>> {
    let mut per_tags: HashMap<&'a str, Vec<&'a liquid::value::Value>> = HashMap::new();
    let add_to_tag =
        |tag: &'a str,
         post: &'a liquid::value::Value,
         per_tags: &mut HashMap<&'a str, Vec<&'a liquid::value::Value>>| {
            let cur_tag = per_tags.entry(tag).or_insert(vec![]);
            cur_tag.push(post)
        };

    for post in all_posts {
        if let Some(tags) = extract_tags(post) {
            for tag in tags {
                let tag = tag.as_scalar().ok_or("Should have string tags")?.to_str();
                add_to_tag(&tag, post, &mut per_tags)
            }
        }
    }
    Ok(per_tags)
}
```

and ended putting `'static` asked by the compiler… is that alright? It feels wrong

Reply:
to avoid `'static` I tried to add `'a` on some functions instead:
```rust
fn create_tags_paginators<'a>(
    all_posts: &[&'a liquid::value::Value],
    doc: &Document,
    pagination_cfg: &PaginationConfig,
) -> Result<Vec<Paginator>> {
```
and 
```rust
pub fn generate_paginators<'a>(
    doc: &mut Document,
    posts_data: &'a [liquid::value::Value],
) -> Result<Vec<Paginator>> {
```

and I get this error:
```rust
error[E0495]: cannot infer an appropriate lifetime for lifetime parameter 'a in function call due to conflicting requirements
   --> src\pagination.rs:278:24
    |
278 |     let mut per_tags = distribute_posts_by_tags(&all_posts)?;
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
note: first, the lifetime cannot outlive the lifetime 'a as defined on the function body at 273:27...
   --> src\pagination.rs:273:27
    |
273 | fn create_tags_paginators<'a>(
    |                           ^^
    = note: ...so that the expression is assignable:
            expected &[&liquid::value::Value]
               found &[&'a liquid::value::Value]
    = note: but, the lifetime must be valid for the static lifetime...
    = note: ...so that the types are compatible:
            expected std::convert::From<&str>
               found std::convert::From<&'static str>
```

Reply:
current `pagination.rs`: https://gist.github.com/Geobert/afa5682b4ee65eb3db74b5d843a845cd

Reply:
It seems that my main issue is either not having a `From<&'a str>` to convert a `&'a str` to a `Value`, or not having a `to_string` that gives `Cow<String>` from a `Value`. (Or maybe I'm totally misunderstanding here ^^')

Reply:
I'm really stuck here, I think it's not possible, shall I keep the non borrowing version?

Reply:
I tried with `&'a String` instead of `&'a str` with no luck: https://gist.github.com/Geobert/71ca84f562e4046d5f7716ae1d1b5e3b

I got:
```rust
  Checking cobalt-bin v0.14.0 (I:\Dev\cobalt.rs)
error[E0308]: mismatched types
   --> src\pagination.rs:254:46
    |
254 |                 let cur_tag = per_tags.entry(tag).or_insert(vec![]);
    |                                              ^^^ expected reference, found enum `std::borrow::Cow`
    |
    = note: expected type `&'a std::string::String`
               found type `std::borrow::Cow<'_, str>`

```

Reply:
Sorry, I don't remember what I was getting at; let's not worry about this for now.